source/application/AMPIC887DriverServer/GCS2Commands/AMGCS2AsyncMoveCommand.cpp

```diff
@@ -0,0 +1,157 @@
+#include "AMGCS2AsyncMoveCommand.h"
+#include "AMGCS2GetMovingStatusCommand.h"
+#include "AMGCS2GetCurrentPositionCommand.h"
+#include "AMGCS2GetTargetPositionCommand.h"
+#include "../AMGCS2Support.h"
+AMGCS2AsyncMoveCommand::AMGCS2AsyncMoveCommand(const AMPIC887AxisMap<double>& targetPositions)
+{
+	command_ = new AMGCS2MoveCommand(targetPositions);
+	targetPositions_ = targetPositions;
+}
+
+AMGCS2AsyncMoveCommand::~AMGCS2AsyncMoveCommand() {
+	delete command_;
+}
+
+AMPIC887AxisMap<double> AMGCS2AsyncMoveCommand::targetPositions() const
+{
+	return targetPositions_;
+}
+
+bool AMGCS2AsyncMoveCommand::validateArguments()
+{
+	if(targetPositions_.isEmpty()) {
+		lastError_ = "No axis positions provided";
+		return false;
+	}
+
+	if(targetPositions_.containsUnknownAxis()) {
+		lastError_ = "Contains unknown axis";
+		return false;
+	}
+
+	return true;
+}
+
+bool AMGCS2AsyncMoveCommand::runImplementation()
+{
+	command_->setController(controller_);
+	command_->run();
+
+	if(command_->runningState() != Succeeded) {
+		lastError_ = command_->lastError();
+		return false;
+	}
+
+	return true;
+}
+
+void AMGCS2AsyncMoveCommand::isFinishedImplementation()
+{
+	AMPIC887AxisCollection axesMoved = targetPositions_.axes();
+
+	// Are we still moving?
+
+	AMGCS2GetMovingStatusCommand movingStatusCommand;
+
+	movingStatusCommand.setController(controller_);
+	movingStatusCommand.run();
+
+	if(movingStatusCommand.runningState() != Succeeded) {
+		lastError_ = "Could not obtain moving status of axes";
+		runningState_ = Failed;
+		return;
+	}
+
+	QFlags<AMGCS2::AxisMovementStatus> movementStatuses =
+			movingStatusCommand.movementStatuses();
+
+	bool allStillMoving = true;
+	foreach(AMGCS2::Axis currentAxis, axesMoved) {
```

Comment:
Just so I know you're sure, if you were doing a move with multiple axes (say x, y, u, v).  This will will become false if any of the axes becomes false but not all of them.  For example, allStillMoving will be false in the situation where x = false, y = true, u = true, and v = true.  From what I can tell from reading the rest of this function, you only want to continue if all the axes have stopped, not when only one has stopped.